src/dataframe/dataframe.jl

```diff
@@ -1090,15 +1090,19 @@ Base.convert(::Type{DataFrame}, A::AbstractMatrix) = DataFrame(A)
 
 Base.convert(::Type{DataFrame}, d::AbstractDict) = DataFrame(d)
 
-
-##############################################################################
-##
-## push! a row onto a DataFrame
-##
-##############################################################################
-
-function Base.push!(df::DataFrame, row::Union{AbstractDict, NamedTuple})
+function Base.push!(df::DataFrame, row::Union{AbstractDict, NamedTuple}; exact::Bool=true)
+    if ncol(df) == 0 && row isa NamedTuple
+        for (n, v) in pairs(row)
+            setproperty!(df, n, fill!(Tables.allocatecolumn(typeof(v), 1), v))
+        end
+        return df
+    end
     i = 1
+    if exact && length(row) != size(df, 2)
```

Comment:
How about using `columns=:equal` and `columns=:intersect` as in https://github.com/JuliaData/DataFrames.jl/pull/1659? Also, it's not just the number of elements that has to match: names must be the same too.

Reply:
OK - fixed

Reply:
> Also, it's not just the number of elements that has to match: names must be the same too.

We do not have to check for this. If names do not match later the code will fail. We could add an additional test here but it will:
* slow down the run time if no error happens;
* speed up the run time if error happens.

(I thought it is better to be fast in no-error case)